drv/lpc55-sprot-server/src/main.rs

```diff
@@ -223,117 +234,181 @@ impl Io {
         }
     }
 
+    // Wait for CSn to actually be de-asserted.
+    //
+    // This checks the actual gpio pin, not the SPI block saturating bit which
+    // may be set from a prior deassertion.
+    //
+    // If CSn is still asserted at this point, then we are desynchronized and so
+    // we return an error.
+    //
+    // See https://github.com/oxidecomputer/hubris/issues/1507 for why we do
+    // this.
+    fn wait_for_csn_deasserted(&mut self) -> Result<(), DesynchronizedError> {
+        let mut result = Ok(());
+        while self.gpio.read_val(CHIP_SELECT) != Value::One {
+            ringbuf_entry!(Trace::Desynchronized);
+            if result.is_ok() {
+                self.stats.desynchronized =
+                    self.stats.desynchronized.wrapping_add(1);
+            }
+            result = Err(DesynchronizedError);
```

Comment:
Nit: maybe move this into the conditional, to make it more obvious that the conditional is only executed on the first transition from `Ok` -> `Err`?

Reply:
good call!